parser/lexer_parser.py

```diff
@@ -0,0 +1,114 @@
+import json
+from dataclasses import dataclass, fields
+from pathlib import Path
+from typing import Any, Callable
+
+parse_fn: Callable[[Path], str] | None = None
+
+
+@dataclass
+class JsonConvertible:
+    @classmethod
+    def is_convertible(cls, json_dict: dict) -> bool:
+        for field in fields(cls):
+            if field.name not in json_dict:
+                return False
+        return True
+
+
+@dataclass
+class FilePosition(JsonConvertible):
+    line: int
+    column: int
+
+
+@dataclass
+class Meta(JsonConvertible):
+    start: FilePosition
+    end: FilePosition
+
+
+@dataclass
+class Token(JsonConvertible):
+    # TODO: rename to token
+    name: str
+    value: str
+    meta: Meta
+
+
+@dataclass
+class Tree(JsonConvertible):
+    name: str
+    children: list[Any]
+    meta: Meta
+
+
+class Transformer:
+    def __init__(self, visit_tokens=False) -> None:
+        self._visit_tokens = visit_tokens
+
+    def transform(self, tree: Tree):
+        my_children = []
+        for child in tree.children:
+            if child is None:
+                my_children.append(None)
+            elif isinstance(child, Token):
+                if self._visit_tokens and hasattr(self, child.name):
+                    my_children.append(getattr(self, child.name)(child))
+                else:
+                    my_children.append(child)
+            else:
+                assert isinstance(child, Tree)
+                my_children.append(self.transform(child))
+
+        new_tree = Tree(tree.name, my_children, tree.meta)
+        if not hasattr(self, tree.name):
+            return new_tree
+
+        return getattr(self, tree.name)(new_tree)
+
+
+class Interpreter:
+    def __init__(self, visit_tokens=False) -> None:
+        self._visit_tokens = visit_tokens
+
+    def visit(self, tree: Tree | Token):
+        if not hasattr(self, tree.name):
+            if isinstance(tree, Token):
+                return tree
+
+            return self.visit_children(tree)
```

Comment:
This deviates slightly from lark's interface

> Unlike Transformer and Visitor, the Interpreter doesn’t automatically visit its sub-branches. The user has to explicitly call visit, visit_children, or use the @visit_children_decor. This allows the user to implement branching and loops.

It's fine if this is intentional

Reply:
Lark actually does the same thing here (despite the docs)... If you run the interpreter and Lark parses a tree that doesn't correspond to a method of your specialized Interpreter, it just visits the children instead (see `__default__()` in the code below)...

Our current `graphene_parser.py` actually requires this feature since we don't actually bother matching the `program` grammar, instead, we're only interested in its top-level children (which lark automatically iterates through).

```python
class Interpreter(_Decoratable, ABC, Generic[_Leaf_T, _Return_T]):
    """Interpreter walks the tree starting at the root.

    Visits the tree, starting with the root and finally the leaves (top-down)

    For each tree node, it calls its methods (provided by user via inheritance) according to ``tree.data``.

    Unlike ``Transformer`` and ``Visitor``, the Interpreter doesn't automatically visit its sub-branches.
    The user has to explicitly call ``visit``, ``visit_children``, or use the ``@visit_children_decor``.
    This allows the user to implement branching and loops.
    """

    def visit(self, tree: Tree[_Leaf_T]) -> _Return_T:
        # There are no guarantees on the type of the value produced by calling a user func for a
        # child will produce. So only annotate the public method and use an internal method when
        # visiting child trees.
        return self._visit_tree(tree)

    def _visit_tree(self, tree: Tree[_Leaf_T]):
        f = getattr(self, tree.data)
        wrapper = getattr(f, 'visit_wrapper', None)
        if wrapper is not None:
            return f.visit_wrapper(f, tree.data, tree.children, tree.meta)
        else:
            return f(tree)

    def visit_children(self, tree: Tree[_Leaf_T]) -> List:
        return [self._visit_tree(child) if isinstance(child, Tree) else child
                for child in tree.children]

    def __getattr__(self, name):
        return self.__default__

    def __default__(self, tree):
        return self.visit_children(tree)
```

Reply:
Ah, I see. yeah that makes sense